/**
 * 
 * @file SetupWidget.h
 * @brief 求解器类型选择界面
 * @author BaGuijun (baguijun@163.com)
 * @date 2024-08-14
 * 
 */
#ifndef _SetupWidget_H
#define _SetupWidget_H

#include "GUICalculateDialogAPI.h"
#include "GUICalculateWidgetBase.h"

namespace Ui {
    class SetupWidget;
}

namespace Interface {
}

namespace EventOper {
    class ParaWidgetInterfaceOperator;
}

namespace GUI
{
    /**
     * @brief 求解器类型选择界面
     * @author BaGuijun (baguijun@163.com)
     * @date 2024-08-14
     */
    class GUICalculateDialogAPI SetupWidget : public GUICalculateWidgetBase
    {
        Q_OBJECT;
    public:
        /**
         * @brief Construct a new Setup Widget object
         * @param[i]  oper           操作器对象
         * @param[i]  parent         父对象
         * @author BaGuijun (baguijun@163.com)
         * @date 2024-08-14
         */
        SetupWidget(EventOper::ParaWidgetInterfaceOperator* oper, QWidget* parent = nullptr);
        /**
         * @brief Destroy the Setup Widget object
         * @author BaGuijun (baguijun@163.com)
         * @date 2024-08-14
         */
        ~SetupWidget();
        /**
         * @brief 初始化
         * @author BaGuijun (baguijun@163.com)
         * @date 2024-08-14
         */
        void init();
        /**
         * @brief 更新表格
         * @author BaGuijun (baguijun@163.com)
         * @date 2024-08-14
         */
        void updateTableWidget();
    private:
        /**
         * @brief 初始化设置类型
         * @author BaGuijun (baguijun@163.com)
         * @date 2024-08-14
         */
        void initSetupType();
        /**
         * @brief 初始化当前的求解器类型
         * @author BaGuijun (baguijun@163.com)
         * @date 2024-08-14
         */
        void initCurrentType();
    private slots:
        ;
        /**
         * @brief选择按钮点击事件
         * @author BaGuijun (baguijun@163.com)
         * @date 2024-08-14
         */
        void on_pushButton_Select_clicked();
        /**
         * @brief求解器类型选择槽函数
         * @author BaGuijun (baguijun@163.com)
         * @date 2024-08-14
         */
        void slotTypeSelect();
    private:
        /**
         * @brief ui对象
         * @author BaGuijun (baguijun@163.com)
         * @date 2024-08-14
         */
        Ui::SetupWidget* _ui = nullptr;
        /**
         * @brief 求解器类型
         * @author BaGuijun (baguijun@163.com)
         * @date 2024-08-14
         */
        QList<QString> _types = {};
    };
}

#endif
